mmengine/optim/optimizer/apex_optimizer_wrapper.py

```diff
@@ -0,0 +1,109 @@
+
+    Warnings:
+        ``ApexOptimWrapper`` requires
+            [nvidia apex](https://github.com/NVIDIA/apex).
+
+    Args:
+
+        **kwargs: Keyword arguments passed to OptimWrapper.
+
+    Note:
+        If you use ``IterBasedRunner`` and enable gradient accumulation,
+        the original `max_iters` should be multiplied by
+        ``accumulative_counts``.
+    """
+
+    def __init__(self, opt_level='O1', loss_scale='dynamic', **kwargs):
+        assert apex_amp is not None, \
+            'Apex is not installed. Please check ' \
+            'https://github.com/NVIDIA/apex#linux.'
+        super().__init__(**kwargs)
+        self.opt_level = opt_level
+        self.loss_scale = loss_scale
+
+    def backward(self, loss: torch.Tensor, **kwargs):
+        """Perform gradient back propagation with :attr:`loss_scaler`.
+
+        Args:
+            loss (torch.Tensor): The loss of current iteration.
+            kwargs: Keyword arguments passed to :meth:`torch.Tensor.backward`
+        """
+        with apex_amp.scale_loss(loss, self.optimizer) as scaled_loss:
+            scaled_loss.backward()
+        self._inner_count += 1
+
+    def state_dict(self) -> dict:
+        """Get the state dictionary of :attr:`optimizer` and
+        :attr:`apex_amp`.
+
+        Based on the state dictionary of the optimizer, the returned state
+        dictionary will add a key named "apex_amp".
+
+        Returns:
+            dict: The merged state dict of :attr:`apex_amp` and
+            :attr:`optimizer`.
+        """
+        state_dict = self.optimizer.state_dict()
+        state_dict['apex_amp'] = apex_amp.state_dict()
+        return state_dict
+
+    def load_state_dict(self, state_dict: dict):
+        """Load and parse the state dictionary of :attr:`optimizer` and
+        :attr:`apex_amp`.
+
+        If state_dict contains "apex_amp", the :attr:`apex_amp` will
+        load the corresponding keys. Otherwise, only the :attr:`optimizer`
+        will load the state dictionary.
+
+        Args:
+            state_dict (dict): The state dict of :attr:`optimizer` and
+                :attr:`apex_amp`
+        """
+        if 'apex_amp' in state_dict:
+            apex_amp.load_state_dict(state_dict.pop('apex_amp'))
+        self.optimizer.load_state_dict(state_dict)
+
+    @contextmanager
+    def optim_context(self, model: nn.Module):
+        """Enables the context for mixed precision training, and enables the
+        context for disabling gradient synchronization during gradient
+        accumulation context.
+
+        Args:
+            model (nn.Module): The training model.
+        """
+        if hasattr(self.optimizer, '_amp_stash'):
+            yield
+        else:
+            if isinstance(model, torch.nn.parallel.DistributedDataParallel):
+                model = model.module
+            with super().optim_context(model):
+                model, self.optimizer = apex_amp.initialize(
+                    model,
```

Comment:
This is just a question. When I checked the Apex documentation, I couldn't find a description to handle it like this.
Isn't there any issue in training when handling it like this?

Reply:
This is actually what I concerned, I took a cursory look at some of implementation for `apex.initialize`:

https://github.com/NVIDIA/apex/blob/082f999a6e18a3d02306e27482cc7486dab71a50/apex/amp/_initialize.py#L145

It seems `initialize` only patches the forward and registers some custom hooks into forward. I'm not sure about doing this on `model.module` will work or not, we need to check the optimization for saving memory and accelerating training when `ApexOptimWrapper` is enabled.

@xcnick could you provide the comparison of `nvidia-smi` when `ApexOptimWrapper` is enabled or not?

Reply:
OK, there is something wrong with current implementation.

In `examples\examples/distributed_training.py`, compare the following two options:
```python
batch_size=1024,
...
optim_wrapper=dict(
        type='ApexOptimWrapper', opt_level='O1', loss_scale=10,
        optimizer=dict(type=Adam, lr=0.001)),
...
```
and

```python
batch_size=1024,
...
optim_wrapper=dict(optimizer=dict(type=SGD, lr=0.001, momentum=0.9)),
```

result:

| optim_wrapper | nvidia-smi | memory in log file |
| -- | -- | -- |
| ApexOptimWrapper |  11GB | 5574 |
| original OptimWrapper | 5.2GB | 2249 |

It seems that calling `amp.initialize` in `optim_context` may not be the correct way.

Reply:
You may be able to participate in the related discussion if you are interested.
https://github.com/open-mmlab/mmengine/discussions/749#discussioncomment-4201408

Reply:
Emmm, the conclusion makes me confused. I've tried to train mmdet with retinanet-r50 based on this PR.  The result of nvidia-smi is:

![image](https://user-images.githubusercontent.com/57566630/204624258-23d98373-36c3-4d7f-813f-b149b2e1907c.png)

![image](https://user-images.githubusercontent.com/57566630/204624280-bda50bc3-abe7-4d28-9dbc-b65347e3005f.png)

It seems that the `ApexOptimWrapper` has worked.... I also test the `examples/examples/distributed_training.py`, the result is the same.

So... what happened 🤣

Reply:
That is so embarrassing.

In addition, I've tested `O0` / `O2` / `O3`:
- `O0`: memory usage is almost the same as the `OptimWrapper`
- `O3`: memory usage is between `O0` and `O1`, and the loss is nan all the time
- `O2`: some compatibility errors occur in the runner

Also, some compatibility modifications are required when apex amp is set to `O0` or `O3`.

Reply:
I'll use `ApexOptimWrapper` based on this PR to train MMDet these days, and check the training speed and accuracy. Do you have any suggestions or comments about the current implementations @xcnick @nijkah?

Reply:
I'm not sure whether the this implementation can work correctly, so good luck ^^

In addition, the `to` method of `BaseModel` in `base_model.py` needs to be modified for compatibility in `O0` and `O3` modes.

Reply:
> In addition, the `to` method of `BaseModel` in `base_model.py` needs to be modified for compatibility in `O0` and `O3` modes.

Do you mean the modifications in #783 ? If it is not enough, feel free to leave a comment!

Reply:
Great job!

Reply:
Sorry for my late reply. To give a conclusion first: the current implementation seems OK. I used `ApexOptimWrapper` to train the `atss` in MMDet, and the loss was able to converge normally, and it will take some time to verify the accuracy.

The reason for the late reply is that I spent some time trying to train RetinaNet (O1) based on `ApexOptimWrapper`. However, it turned out that neither `AmpOptimWrapper` nor `ApexOptimWrapper· can train RetinaNet (loss is `nan`) normally, which is determined by the model's own characteristics and has nothing to do with the implementation of `ApexOptimWrapper`